presto-main/src/main/java/com/facebook/presto/memory/HighMemoryTaskKiller.java

```diff
@@ -0,0 +1,248 @@
+            if (isLowMemory() && beforeGcDataSize.toBytes() - afterGcDataSize.toBytes() < reclaimMemoryThreshold) {
+                triggerTaskKiller = true;
+            }
+        }
+
+        return triggerTaskKiller;
+    }
+
+    private List<SqlTask> getActiveTasks()
+    {
+        return sqlTaskManager.getAllTasks().stream()
+                .filter(task -> !task.getTaskState().isDone())
+                .collect(toImmutableList());
+    }
+
+    @VisibleForTesting
+    public static Optional<QueryId> getMaxMemoryConsumingQuery(ListMultimap<QueryId, SqlTask> queryIDToSqlTaskMap)
+    {
+        if (queryIDToSqlTaskMap.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Comparator<Map.Entry<QueryId, Long>> comparator = Comparator.comparingLong(Map.Entry::getValue);
+
+        Optional<QueryId> maxMemoryConsumpingQueryId = queryIDToSqlTaskMap.asMap().entrySet().stream()
+                //Convert to Entry<QueryId, Long>, QueryId -> Total Memory Reservation
+                .map(entry ->
+                        new AbstractMap.SimpleEntry<>(entry.getKey(), entry.getValue().stream()
+                                .map(SqlTask::getTaskInfo)
+                                .map(TaskInfo::getStats)
+                                .mapToLong(stats -> stats.getUserMemoryReservationInBytes() + stats.getSystemMemoryReservationInBytes() + stats.getRevocableMemoryReservationInBytes())
+                                .sum())
+                ).max(comparator.reversed()).map(Map.Entry::getKey);
+
+        return maxMemoryConsumpingQueryId;
+    }
+
+    private boolean isFrequentFullGC(long lastFullGCTimestamp, long currentFullGCTimestamp)
+    {
+        long diffBetweenFullGCMilis = (currentFullGCTimestamp - lastFullGCTimestamp) / 1000000;
+        log.debug("Time difference between last 2 full GC in miliseconds: " + diffBetweenFullGCMilis);
+        if (diffBetweenFullGCMilis > taskKillerFrequentFullGCDurationThreshold.getValue(TimeUnit.MILLISECONDS)) {
+            log.debug("Skip killing tasks Due to full GCs were not happening frequently.");
+            return false;
+        }
+        return true;
+    }
+
+    private boolean hasFullGCFreedEnoughBytes(long currentGarbageCollectedBytes)
+    {
+        if (currentGarbageCollectedBytes < reclaimMemoryThreshold && lastFullGCCollectedBytes < reclaimMemoryThreshold) {
+            log.debug("Full GC not able to free enough memory. Current freed bytes: " + currentGarbageCollectedBytes + " previously freed bytes: " + lastFullGCCollectedBytes);
+            return false;
+        }
+        log.debug("Full GC able to free enough memory. Current freed bytes: " + currentGarbageCollectedBytes + " previously freed bytes: " + lastFullGCCollectedBytes);
+        return true;
+    }
+
+    private boolean isLowMemory()
+    {
+        MemoryUsage memoryUsage = memoryMXBean.getHeapMemoryUsage();
+
+        if (memoryUsage.getUsed() > heapMemoryThreshold) {
```

Comment:
Do we really need this helper method ? If so, how about return (memoryUsage.getUsed() > heapMemThresh) ?